h/static/scripts/directive/annotation.js

```diff
@@ -0,0 +1,730 @@
+          angular.copy(updatedModel, model);
+          $rootScope.$emit('annotationUpdated', model);
+          vm.view();
+        };
+        onRejected = function(reason) {
+          flash.error(
+            errorMessage(reason), 'Saving annotation failed');
+        };
+        return updatedModel.$update({
+          id: updatedModel.id
+        }).then(onFulfilled, onRejected);
+    }
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#reply
+    * @description
+    * Creates a new message in reply to this annotation.
+    */
+  vm.reply = function() {
+    var id = model.id;
+    var references = model.references || [];
+    var reply;
+    var uri = model.uri;
+
+    if (typeof references === 'string') {
+      references = [references];
+    }
+
+    references = references.concat(id);
+
+    reply = annotationMapper.createAnnotation({
+      references: references,
+      uri: uri
+    });
+    reply.group = model.group;
+
+    if (session.state.userid) {
+      if (permissions.isShared(model.permissions, model.group)) {
+        reply.permissions = permissions.shared(reply.group);
+      } else {
+        reply.permissions = permissions.private();
+      }
+    }
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#render
+    * @description Called to update the view when the model changes.
+    */
+  vm.render = function() {
+    var documentTitle;
+    var domain;
+    var draft = drafts.get(model);
+    var i;
+    var link;
+    var tagsAsObjects;
+    var uri = model.uri;
+
+    // Extend the view model with a copy of the domain model.
+    // Note that copy is used so that deep properties aren't shared.
+    vm.annotation = angular.extend({}, angular.copy(model));
+
+    // If we have unsaved changes to this annotation, apply them
+    // to the view model.
+    if (draft) {
+      angular.extend(vm.annotation, angular.copy(draft));
+    }
+
+    vm.annotationURI = new URL(
+      '/a/' + vm.annotation.id, vm.baseURI).href;
+
+    // Extract the document metadata.
+    domain = new URL(uri).hostname;
+    if (model.document) {
+      if (uri.indexOf('urn') === 0) {
+        for (i = 0; i < model.document.link.length; i++) {
+          link = model.document.link[i];
+          if (!(link.href.indexOf('urn'))) {
+            continue;
+          }
+          uri = link.href;
+          break;
+        }
+      }
+
+      documentTitle = Array.isArray(
+        model.document.title) ? model.document.title[0] : model.document.title;
+
+      vm.document = {
+        uri: uri,
+        domain: domain,
+        title: documentTitle || domain
+      };
+    } else {
+      vm.document = {
+        uri: uri,
+        domain: domain,
+        title: domain
+      };
+    }
+
+    if (vm.document.title.length > 30) {
+      vm.document.title = vm.document.title.slice(0, 30) + '…';
+    }
+
+    // Form the tags for ngTagsInput.
+    tagsAsObjects = [];
```

Comment:
This could be written more simply as:
```js
vm.annotation.tags = (vm.annotation.tags || []).map(function (tag) {
  return { text: tag };
});
```

Its a bit confusing though that we first set 'annotation.tags' to a string array and then change it to an array of `{ text: string }`